README.md
# RSMatcher

RSMatcher is a custom application written for a non-profit educational program that matches volunteer reading coaches with children in need of reading assistance.
The coaches meet and read with each student for 30 minutes each week.
There may be hundreds of coaches and over one thousand students.
RSMatcher is an implementation of an algorithm to find a near-optimal scheduling solution.
An optimal solution matches as many students as coaches are willing to take while also meeting the scheduling restrictions of the coaches and schools.
The problem space is first converted to a graph in which finding all cycles identifies possible schedule additions.
Heuristic sorting is used to prioritize more efficient additions (e.g. a coach meets with multiple students at the same school back-to-back).
At multiple points, shuffling and round-robin interleaving is used to remove bias.
Then a branch and bound traversal of the solution tree is used to converge on a near-optimal solution while also culling solutions with relatively poor scores.


CHANGELOG.md
# Changelog

This document contains high-level descriptions of the changes made to
the software in each released version.

## 2019/08/26 v4
- Detect invalid referral entries that select times for a day and NONE
- Write invalid input entries to output terminal
- Write invalid input entries to files in same directory as database

## 2019/09/16 v5
- Fix off-by-one error that was not picking up 2nd recess
- Extend default end time to 15:30 and fix bug in error path
- Add more useful error text for referrals without schedules and fix
  problem with referrals that use all possible student inputs
- Add manual assignments and assignment checks against schedules

## 2019/09/18 v6
- Change how pre-existing assignments are applied to the database and
  used in the matcher

## 2021/08/29 v7
- Update to use 2021 CSV columns
- Last column of assignments CSV is now either a timestamp or "manual"
- Build script to release as single EXE file.

build.py

import os
import os.path
import shutil

import markdown
import PyInstaller.__main__

doc_template = """
<html>
<head>
<style>
body {
    font-family: Helvetica, Arial, Sans-Serif;
    margin: auto;
    width: 40em;
}
</style>
</head>
<body>
%s
</body>
"""

def main(): 
    if os.path.exists('dist'):
        shutil.rmtree('dist')
    
    os.mkdir('dist')
    os.mkdir('dist/RSMatcher')
   
    with open('RSMatcher.md', 'r') as fin:
        with open('dist/RSMatcher/RSMatcher.html', 'w') as fout:
            body = markdown.markdown(
                text=fin.read(),
                output_format='html5',
            )
            fout.write(doc_template % body)

    with open('CHANGELOG.md', 'r') as fin:
        with open('dist/RSMatcher/CHANGELOG.html', 'w') as fout:
            body = markdown.markdown(
                text=fin.read(),
                output_format='html5',
            )
            fout.write(doc_template % body)
        
    if os.path.exists('build'):
        shutil.rmtree('build')
    
    if os.path.exists('RSMatcher.spec'):
        os.unlink('RSMatcher.spec')

    PyInstaller.__main__.run([
        'RSMatcher.py',
        '--distpath',
        './dist/RSMatcher/',
        '--onefile',
    ])


if __name__ == '__main__':
    main()

RSMatcher.py
"""
Copyright (c) 2019 Mark Siner

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from rsmatch import gui


if __name__ == '__main__':
    gui.main()

rsmatch/gui.py
"""
Copyright (c) 2019 Mark Siner

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""


import os.path
import threading
from subprocess import Popen
import tkinter as tk
from tkinter import filedialog

from . import input
from . import output
from . import matcher
from . import database


def ask_for_file(string_var):
    initial_dir = None
    if string_var.get():
        initial_dir = os.path.dirname(string_var.get())
    string_var.set(
        filedialog.askopenfilename(
            title="Select file",
            filetypes=(('CSV files', '*.csv'), ('all files','*.*')),
            initialdir=initial_dir))

def ask_for_dir(string_var):
    initial_dir = None
    if string_var.get():
        initial_dir = string_var.get()
    string_var.set(
        filedialog.askdirectory(
            title="Select Directory",
            mustexist=True,
            initialdir=initial_dir))

class MatcherGUI(tk.Tk):
    def __init__(self):
        super().__init__()

        self.referrals_path = tk.StringVar()
        self.coaches_path = tk.StringVar()
        self.assignments_path = tk.StringVar()
        self.output_dir_path = tk.StringVar()
        self.first_choice = tk.IntVar()
        self.first_choice.set(True)
        self.second_choice = tk.IntVar()
        self.greatest_need = tk.IntVar()
        self.database_path = None
        self.btn_database = None
        self.btn_report = None
        self.lst_school = None
        self.lst_school_data = []
        self.btn_start = None

        self.title('RS Matcher')

        tk.Label(
            self,
            text='STEP 1: Select teacher referral CSV file input').pack(
                anchor=tk.W)

        frame = tk.Frame()
        frame.pack(anchor=tk.W)
        self.referral_entry = tk.Entry(
            frame, width=50, textvariable=self.referrals_path)
        self.referral_entry.grid(
            row=0, column=0, columnspan=4, padx=2, pady=2, sticky=tk.W)
        tk.Button(
            frame,
            text='Browse',
            command=lambda: ask_for_file(self.referrals_path)).grid(
                row=0, column=4, padx=2, pady=2, sticky=tk.W)

        tk.Label(
            self,
            text='STEP 2: Select coach registration CSV file input').pack(
                anchor=tk.W)

        frame = tk.Frame()
        frame.pack(anchor=tk.W)
        self.referral_entry = tk.Entry(
            frame, width=50, textvariable=self.coaches_path)
        self.referral_entry.grid(
            row=0, column=0, columnspan=4, padx=2, pady=2, sticky=tk.W)
        tk.Button(
            frame,
            text='Browse',
            command=lambda: ask_for_file(self.coaches_path)).grid(
                row=0, column=4, padx=2, pady=2, sticky=tk.W)

        tk.Label(
            self,
            text='STEP 3 (optional): Select assignments CSV file input').pack(
                anchor=tk.W)

        frame = tk.Frame()
        frame.pack(anchor=tk.W)
        self.referral_entry = tk.Entry(
            frame, width=50, textvariable=self.assignments_path)
        self.referral_entry.grid(
            row=0, column=0, columnspan=4, padx=2, pady=2, sticky=tk.W)
        tk.Button(
            frame,
            text='Browse',
            command=lambda: ask_for_file(self.assignments_path)).grid(
                row=0, column=4, padx=2, pady=2, sticky=tk.W)

        tk.Label(
            self,
            text='STEP 4: Select folder for saving output').pack(
                anchor=tk.W)

        frame = tk.Frame()
        frame.pack(anchor=tk.W)
        self.referral_entry = tk.Entry(
            frame, width=50, textvariable=self.output_dir_path)
        self.referral_entry.grid(
            row=0, column=0, columnspan=4, padx=2, pady=2, sticky=tk.W)
        tk.Button(
            frame,
            text='Browse',
            command=lambda: ask_for_dir(self.output_dir_path)).grid(
                row=0, column=4, padx=2, pady=2, sticky=tk.W)
                
        tk.Label(self, text='STEP 5: Create database from input files').pack(
            anchor=tk.W, padx=2, pady=2)
        frame = tk.Frame()
        frame.pack(anchor=tk.W)
        self.btn_database = tk.Button(
            frame,
            text='Create Database',
            command=self.create_database)
        self.btn_database.grid(row=0, column=0, padx=2, pady=2)
        self.btn_report = tk.Button(
            frame,
            text='Create Report',
            state=tk.DISABLED,
            command=self.create_report)
        self.btn_report.grid(row=0, column=1, padx=2, pady=2)
        
        tk.Label(self, text='STEP 6: Select a school for matching').pack(anchor=tk.W)
        self.lst_school = tk.Listbox(self, selectmode=tk.SINGLE)
        self.lst_school.pack(anchor=tk.W, padx=2, pady=2)

        tk.Label(self, text='STEP 7: Select school preferences to use').pack(anchor=tk.W)
        frame = tk.Frame()
        frame.pack(anchor=tk.W)
        tk.Checkbutton(frame, text='First', variable=self.first_choice).grid(
            row=0, column=0, columnspan=1, padx=2, pady=2, sticky=tk.W)
        tk.Checkbutton(frame, text='Second', variable=self.second_choice).grid(
            row=0, column=1, columnspan=1, padx=2, pady=2, sticky=tk.W)
        tk.Checkbutton(frame, text='Greatest Need', variable=self.greatest_need).grid(
            row=0, column=2, columnspan=1, padx=2, pady=2, sticky=tk.W)

        tk.Label(self, text='STEP 8: Start matching algorithm').pack(anchor=tk.W)
        self.btn_start = tk.Button(
            self,
            text='Start',
            state=tk.DISABLED,
            command=self.run)
        self.btn_start.pack(anchor=tk.W, padx=2, pady=2)

    def create_database(self):
        referrals_path = self.referrals_path.get()
        coaches_path = self.coaches_path.get()
        assignments_path = self.assignments_path.get()
        output_dir_path = self.output_dir_path.get()
        output_path = os.path.join(output_dir_path, 'rsdb.json')
        self.database_path = None
        input.create_database(
            referrals_path, coaches_path, assignments_path, output_path)
        self.database_path = output_path
        self.btn_database.config(state=tk.DISABLED)
        
        rsdb = database.RSDatabase.from_path(output_path)
        
        self.lst_school.delete(0, tk.END)
        self.lst_school_data = []
        for school in sorted(rsdb.schools, key=lambda x: x.name):
            self.lst_school.insert(tk.END, school.name)
            self.lst_school_data.append(school.name)
        self.lst_school.config(state=tk.NORMAL)
        if self.lst_school_data:
            self.lst_school.select_set(0)
        self.btn_report.config(state=tk.NORMAL)
        self.btn_start.config(state=tk.NORMAL)

    def create_report(self):
        self.btn_report.config(state=tk.DISABLED)
        output_dir_path = self.output_dir_path.get()
        rsdb = database.RSDatabase.from_path(self.database_path)
        out = output.MatcherOutput(rsdb, dir=output_dir_path)
        output_path = os.path.join(output_dir_path, 'report.txt')
        out.create_report(output_path)
        self.btn_report.config(state=tk.NORMAL)
        Popen(['notepad.exe', output_path])
        
    def run(self):
        self.btn_start.config(state=tk.DISABLED)
        def runner():
            try:
                school_name = self.lst_school_data[
                    self.lst_school.curselection()[0]]
                matcher.do_match(
                    self.database_path,
                    school=school_name,
                    first=self.first_choice.get(),
                    second=self.second_choice.get(),
                    greatest=self.greatest_need.get())
                output_dir_path = self.output_dir_path.get()
                rsdb = database.RSDatabase.from_path(self.database_path)
                db_output = output.MatcherOutput(rsdb, dir=output_dir_path)
                db_output.create_assignments_csv()
            except Exception as ex:
                print('ERROR: %s' % ex)
            self.btn_start.config(state=tk.NORMAL)
        run_thread = threading.Thread(target=runner)
        run_thread.start()


def main():
    gui = MatcherGUI()
    gui.mainloop()


if __name__ == '__main__':
    main()

rsmatch/database.py
"""
Copyright (c) 2019 Mark Siner

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""


import math
import json
import datetime
import collections


class RSDatabase(object):
    """This class is a data structure for all of the assignment data.
    It is designed to be dumped to and loaded from a single file.
    """

    def __init__(self):
        self.path = None
        self.schools = []
        self.coaches = []
        self.catalog = Catalog()
        self.metadata = Metadata()
        self._assignments = []

    @property
    def assignments(self):
        return self._assignments.copy()

    @classmethod
    def from_path(cls, path):
        database = cls()
        database.path = path
        with open(path, 'r') as json_file:
            json_obj = json.load(json_file)
        schools = [School.from_json_obj(x) for x in json_obj['schools']]
        database.schools = schools
        coaches = [Coach.from_json_obj(x) for x in json_obj['coaches']]
        database.coaches = coaches
        database.catalog = Catalog.from_json_obj(json_obj['catalog'])
        database.metadata = Metadata.from_json_obj(json_obj['metadata'])
        database.init_catalog()
        for x in json_obj['assignments']:
            assign = tuple(x['assign'])
            manual = bool(x['manual'])
            timestamp = x['timestamp']
            if not manual:
                database.check_assignment(assign)
            database.add_assignment(assign, manual=manual, timestamp=timestamp)
        return database

    def save(self, path=None):
        if path is None:
            path = self.path
        if path is None:
            raise ValueError('No path specified')
        json_obj = collections.OrderedDict([
            ('metadata', self.metadata.to_json_obj()),
            ('assignments', self._assignments),
            ('schools', [x.to_json_obj() for x in self.schools]),
            ('coaches', [x.to_json_obj() for x in self.coaches]),
            ('catalog', self.catalog.to_json_obj())])
        with open(path, 'w') as json_file:
            json.dump(json_obj, json_file, indent=2)
        self.path = path

    def init_catalog(self):
        self.catalog.reset()
        for school in self.schools:
            for student in school.students:
                self.catalog.assign_guid(student)
            for teacher in school.teachers:
                self.catalog.assign_guid(teacher)
        for coach in self.coaches:
            coach.guid = self.catalog.assign_guid(coach)

    def find_school(self, teacher_or_student):
        for school in self.schools:
            for teacher in school.teachers:
                if teacher_or_student == teacher:
                    return school
            for student in school.teachers:
                if teacher_or_student == student:
                    return school
        return None

    def add_assignment(self, assign, manual=False, timestamp=None):
        day, slot, _, s_guid, c_guid = assign
        student = self.catalog.get_obj(s_guid)
        coach = self.catalog.get_obj(c_guid)
        
        student.assigned = True
        coach.assigned_days.add(day)
        coach.assignments.add(assign)
        for slot_i in range(self.metadata.slots_per_assignment):
            coach.schedule[day][slot + slot_i] = 2
            student.schedule[day][slot + slot_i] = 2
            
        index = 0
        for assign_i in range(len(self._assignments)):
            curr_assign = self._assignments[assign_i]['assign']
            if curr_assign < assign:
                index = assign_i + 1
            elif curr_assign == assign:
                return
            elif curr_assign > assign:
                break
        self._assignments.insert(index, {
            'assign': assign,
            'manual': manual,
            'timestamp': timestamp,
        })
        
    def check_assignment(self, assign):
        day, slot, t_guid, s_guid, c_guid = assign
        teacher = self.catalog.get_obj(t_guid)
        student = self.catalog.get_obj(s_guid)
        coach = self.catalog.get_obj(c_guid)
        school = self.find_school(teacher)
        day_str = self.metadata.day_to_str(day)
        time_str = self.metadata.slot_to_time(slot)
        num_slots = self.metadata.slots_per_assignment
        for slot_val in coach.schedule[day][slot:slot + num_slots]:
            if not slot_val:
                raise ValueError(
                    'Coach not available %s at %s' % (day_str, time_str))
        for slot_val in student.schedule[day][slot:slot + num_slots]:
            if not slot_val:
                raise ValueError(
                    'Student not available %s at %s' % (day_str, time_str))
        for curr in self._assignments:
            curr_assign = curr['assign']
            curr_day, curr_slot, _, curr_s, curr_c = curr_assign
            curr_school = self.find_school(teacher)
            if curr_assign == assign:
                raise ValueError('Duplicate assignment')
            if curr_s == s_guid:
                raise ValueError('Student already assigned %s %s %s' % (student.to_json_obj(), assign, curr_assign))
            if (curr_day, curr_slot, curr_c) == (day, slot, c_guid):
                raise ValueError(
                    'Coach already assigned on %s at %s' %
                    (day_str, time_str))
            if (curr_day, curr_c) == (day, c_guid) and curr_school != school:
                raise ValueError(
                    'Coach already assigned to a different school on %s' %
                    day_str)


class Metadata(object):

    DAYS = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']

    def __init__(self):
        self.minutes_per_slot = 15
        self.minutes_per_assignment = 30
        self.start_time = datetime.timedelta(hours=8, minutes=30)
        self.end_time = datetime.timedelta(hours=15, minutes=30)
        self.slots_per_assignment = int(
            math.ceil(self.minutes_per_assignment / self.minutes_per_slot))

    def time_to_slot(self, time_str):
        hours, minutes = time_str.split(':')
        curr_time = datetime.timedelta(hours=int(hours), minutes=int(minutes))
        if curr_time < self.start_time:
            curr_time = curr_time + datetime.timedelta(hours=12)
        if curr_time > self.end_time:
            raise ValueError(
                'Time falls outside window %s' % curr_time)
        delta_m = (curr_time - self.start_time).total_seconds() // 60
        return int(delta_m // self.minutes_per_slot)

    def slot_to_time(self, slot):
        delta = datetime.timedelta(minutes=self.minutes_per_slot * slot)
        slot_time = self.start_time + delta
        seconds = slot_time.total_seconds()
        hours = int(seconds // (60 * 60))
        seconds -= hours * (60 * 60)
        minutes = int(seconds // 60)
        #am_pm = 'AM'
        if hours > 12:
            hours -= 12
            #am_pm = 'PM'
        return '%d:%02d' % (hours, minutes)
    
    def day_to_str(self, day):
        return self.DAYS[day]

    def str_to_day(self, day_str):
        return self.DAYS.index(day_str)

    @property
    def num_slots(self):
        total_m = (self.end_time - self.start_time).total_seconds() // 60
        return int(total_m // self.minutes_per_slot)

    @property
    def num_days(self):
        return 5

    def to_json_obj(self):
        times = []
        for time_dt in (self.start_time, self.end_time):
            seconds = time_dt.total_seconds()
            hours = int(seconds // (60 * 60))
            seconds -= hours * (60 * 60)
            minutes = int(seconds // 60)
            times.append('%02d:%02d' % (hours, minutes))
        return collections.OrderedDict([
            ('minutes_per_slot', self.minutes_per_slot),
            ('minutes_per_assignment', self.minutes_per_assignment),
            ('start_time', times[0]),
            ('end_time', times[1])])
    
    @classmethod
    def from_json_obj(cls, json_obj):
        metadata = cls()
        metadata.minutes_per_slot = int(json_obj['minutes_per_slot'])
        metadata.minutes_per_assignment = int(
            json_obj['minutes_per_assignment'])
        times = []
        for time_str in (json_obj['start_time'], json_obj['end_time']):
            hours, mins = time_str.split(':')
            times.append(datetime.timedelta(
                hours=int(hours), minutes=int(mins)))
        metadata.start_time = times[0]
        metadata.end_time = times[1]
        metadata.slots_per_assignment = int(math.ceil(
            metadata.minutes_per_assignment / metadata.minutes_per_slot))
        return metadata


class Catalog(object):
    
    def __init__(self):
        self.obj_to_guid = {}
        self.guid_to_obj = {}
        self.curr_student = 100000
        self.curr_coach = 200000
        self.curr_teacher = 300000

    def reset(self):
        self.obj_to_guid = {}
        self.guid_to_obj = {}

    def assign_guid(self, obj):
        obj_guid = obj.guid
        if obj_guid is None:
            if obj in self.obj_to_guid:
                obj_guid = self.obj_to_guid[obj]
            elif isinstance(obj, Student):
                obj_guid = self.curr_student
                self.curr_student += 1
            elif isinstance(obj, Coach):
                obj_guid = self.curr_coach
                self.curr_coach += 1
            elif isinstance(obj, Teacher):
                obj_guid = self.curr_teacher
                self.curr_teacher += 1
        obj.guid = obj_guid
        self.obj_to_guid[obj] = obj_guid
        self.guid_to_obj[obj_guid] = obj
        return obj_guid

    def get_guid(self, obj):
        return self.obj_to_guid[obj]

    def get_obj(self, obj_guid):
        return self.guid_to_obj[obj_guid]

    def to_json_obj(self):
        return collections.OrderedDict([
            ('curr_student', self.curr_student),
            ('curr_coach', self.curr_coach),
            ('curr_teacher', self.curr_teacher)])

    @classmethod
    def from_json_obj(cls, json_obj):
        catalog = cls()
        catalog.curr_student = json_obj['curr_student']
        catalog.curr_coach = json_obj['curr_coach']
        catalog.curr_other = json_obj['curr_teacher']
        return catalog


class School(object):

    def __init__(self):
        self.name = None
        self.students = []
        self.teachers = []

    def to_json_obj(self):
        return collections.OrderedDict([
            ('name', self.name),
            ('students', [x.to_json_obj() for x in self.students]),
            ('teachers', [x.to_json_obj() for x in self.teachers])])

    @classmethod
    def from_json_obj(cls, json_obj):
        school = cls()
        school.name = json_obj['name']

        json_students = json_obj['students']
        students = [Student.from_json_obj(x) for x in json_students]
        school.students = students

        json_teachers = json_obj['teachers']
        teachers = [Teacher.from_json_obj(x) for x in json_teachers]
        school.teachers = teachers
        return school


class Teacher(object):

    def __init__(self):
        self.guid = None
        self.email = None
        self.first = None
        self.last = None
        self.grade = None

    def to_json_obj(self):
        return collections.OrderedDict([
            ('guid', self.guid),
            ('email', self.email),
            ('first', self.first),
            ('last', self.last),
            ('grade', self.grade)])

    @classmethod
    def from_json_obj(cls, json_obj):
        teacher = cls()
        teacher.guid = json_obj['guid']
        teacher.email = json_obj['email']
        teacher.first = json_obj['first']
        teacher.last = json_obj['last']
        teacher.grade = json_obj['grade']
        return teacher


class Student(object):

    def __init__(self):
        self.guid = None
        self.student_id = None
        self.teacher = None
        self.first = None
        self.last = None
        self.grade = None
        self.gender = None
        self.ell = None
        self.schedule = None
        self.assigned = False

    def to_json_obj(self):
        return collections.OrderedDict([
            ('guid', self.guid),
            ('student_id', self.student_id),
            ('teacher', self.teacher),
            ('first', self.first),
            ('last', self.last),
            ('grade', self.grade),
            ('gender', self.gender),
            ('ell', self.ell),
            ('schedule', self.schedule)])

    @classmethod
    def from_json_obj(cls, json_obj):
        student = cls()
        student.guid = json_obj['guid']
        student.student_id = json_obj['student_id']
        student.teacher = json_obj['teacher']
        student.first = json_obj['first']
        student.last = json_obj['last']
        student.grade = json_obj['grade']
        student.gender = json_obj['gender']
        student.ell = json_obj['ell']
        student.schedule = json_obj['schedule']
        return student


class Coach(object):

    def __init__(self):
        self.guid = None
        self.vid = None
        self.first = None
        self.last = None
        self.schools = None
        self.grades = None
        self.num_students = None
        self.num_days = None
        self.schedule = None
        self.assigned_days = set()
        self.assignments = set()

    def to_json_obj(self):
        return collections.OrderedDict([
            ('guid', self.guid),
            ('vid', self.vid),
            ('email', self.email),
            ('first', self.first),
            ('last', self.last),
            ('schools', self.schools),
            ('grades', self.grades),
            ('num_days', self.num_days),
            ('num_students', self.num_students),
            ('schedule', self.schedule)])

    @classmethod
    def from_json_obj(cls, json_obj):
        coach = cls()
        coach.guid = json_obj['guid']
        coach.vid = json_obj['vid']
        coach.email = json_obj['email']
        coach.first = json_obj['first']
        coach.last = json_obj['last']
        coach.schools = json_obj['schools']
        coach.grades = json_obj['grades']
        coach.num_students = json_obj['num_students']
        coach.num_days = json_obj['num_days']
        coach.schedule = json_obj['schedule']
        return coach